strawberryfields/apps/qchem/dynamics.py

```diff
@@ -117,3 +118,80 @@ def op(t, Ul, w, q):
         sf.ops.Interferometer(Ul) | q
 
     return op
+
+
+def sample_fock(
+    input_state: list,
+    t: float,
+    Ul: np.ndarray,
+    w: np.ndarray,
+    n_samples: int = 1,
+    loss: float = 0.0,
+    cutoff: int = 5,
+) -> list:
+
+    r"""Generate samples for simulating vibrational quantum dynamics with a Fock input state.
+
+    **Example usage:**
+    >>> input_state = [0, 2]
+    >>> t = 10.0
+    >>> Ul = np.array([[0.707106781, -0.707106781],
+    >>>                [0.707106781, 0.707106781]])
+    >>> w = np.array([3914.92, 3787.59])
+    >>> n_samples = 5
+    >>> sample_fock(input_state, t, Ul, w, n_samples = n_samples)
+    [[0, 2], [0, 2], [1, 1], [0, 2], [0, 2]]
+
+    Args:
+        input_state (list): input Fock state
+        t (float): time in femtoseconds
+        Ul (array): normal to local transformation matrix
+        w (array): normal mode frequencies :math:`\omega` in units of :math:`\mbox{cm}^{-1}`
+        n_samples (int): number of samples to be generated
+        loss (float): loss parameter denoting the fraction of generated photons that are lost
+        cutoff (int): cutoff dimension for determining the set of number states for each mode
+
+    Returns:
+        list[list[int]]: a list of samples
+    """
+    if n_samples < 1:
+        raise ValueError("Number of samples must be at least one")
+
+    if t < 0:
+        raise ValueError("Time must be zero or positive")
+
+    if np.any(w <= 0):
+        raise ValueError("Vibrational frequencies must be larger than zero")
+
+    if np.any(np.iscomplex(Ul)):
+        raise ValueError("The normal mode to local mode transformation matrix must be real")
+
+    if not 0 <= loss <= 1:
+        raise ValueError("Loss parameter must take a value between zero and one")
+
+    modes = len(Ul)
+    op = evolution(modes)
+    s = []
+
+    eng = sf.Engine("fock", backend_options={"cutoff_dim": cutoff})
+
+    gbs = sf.Program(modes)
+
+    # pylint: disable=expression-not-assigned
+    with gbs.context as q:
+
+        for i in range(modes):
+            sf.ops.Fock(input_state[i]) | q[i]
+
+        op(t, Ul, w) | q
+
+        if loss:
+            for _q in q:
+                sf.ops.LossChannel(1 - loss) | _q
+
+        sf.ops.MeasureFock() | q
+
+        for _ in range(n_samples):
+            s.append(eng.run(gbs).samples[0].tolist())
```

Comment:
I don't think this should be within the `gbs.context` (i.e., we need to lose a tab)

Also, we should be able to just do:
`eng.run(prog, shots=n_samples).samples.tolist()` (see for example `apps.sample.sample`).

Reply:
Using `shots=2` gives this error: "The operation MeasureFock has not been implemented in FockBackend for the arguments {'shots': 2}"